In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.discrete.discrete_model import Logit, Probit
from __future__ import division
import seaborn as sns

%matplotlib inline
import matplotlib.pylab as plt
from matplotlib.pylab import rcParams
rcParams['figure.figsize'] = 18.5, 10.5

def print_resids(preds, resids):
    ax = sns.regplot(preds, resids);
    ax.set(xlabel = 'Predicted values', ylabel = 'errors', title = 'Predicted values vs. Errors')
    plt.show();

## Problem 2

data on districts from on the 105th Congress (1997-8)

In [2]:
congress = pd.read_stata("http://rlhick.people.wm.edu/econ407/data/congressional_105.dta" )

In [3]:
congress.head()

state  fipstate  sc  cd  repub  age65   black  blucllr  city  coast  \
0    AK         2  81   1    1.0  25898   22566    31560     0      1   
1    AL         1  41   1    1.0  72534  164556    43449     0      1   
2    AL         1  41   2    1.0  75396  139311    49886     0      0   
3    AL         1  41   3    1.0  74506  150175    65849     0      0   
4    AL         1  41   4    1.0  84691   38087    74068     0      0   

    ...     miltpop  nucplant    popsqmi  populatn  rurlfarm  transprt  \
0   ...       24991         0   0.964357    629099      1292     20903   
1   ...        1524         0  85.095802    577630      5475     10925   
2   ...       11250         1  56.973942    577203     12491     11820   
3   ...        5804         0  66.190620    577116      8863      7891   
4   ...         463         0  63.142357    577058     16664      9711   

   unemplyd      union   urban  whlretl  
0     26234  30.400000  222119    50986  
1     19799  18.200001  301197    52983  
2     14127  18.200001  202022    52623  
3     17303  18.200001  157275    45284  
4     16724  18.200001   71451    48266  

[5 rows x 32 columns]

### (a)

For the logit, probit and OLS (linear probability) models, I will estimate

$$ repub = \beta_0 + \beta_1 per\_age65 + \beta_2 per\_black + \beta_3 per\_bluecllr + \beta_4 city + \beta_5 mdnincm + \beta_6 per\_unemployed + \beta_7 union $$


where per_unemployed is the percentage of the district that is unemployed.

In [4]:

congress['per_unemployed'] = congress[['unemplyd', 'populatn']].apply(lambda row: row[0]/ row[1], axis = 1)

In [5]:
#making other percentages
variables = [ 'age65', 'black', 'blucllr']

for v in variables:
    congress['per_' + v] = congress[[v, 'populatn']].apply(lambda row: row[0] / row[1], axis = 1)

In [6]:
congress.columns.values

array(['state', 'fipstate', 'sc', 'cd', 'repub', 'age65', 'black',
       'blucllr', 'city', 'coast', 'construc', 'cvllbrfr', 'enroll',
       'farmer', 'finance', 'forborn', 'gvtwrkr', 'intrland', 'landsqmi',
       'mdnincm', 'miltinst', 'miltmajr', 'miltpop', 'nucplant', 'popsqmi',
       'populatn', 'rurlfarm', 'transprt', 'unemplyd', 'union', 'urban',
       'whlretl', 'per_unemployed', 'per_age65', 'per_black', 'per_blucllr'], dtype=object)

In [7]:
#check for null values
congress[congress.repub.isnull()]

state  fipstate  sc  cd  repub   age65  black  blucllr  city  coast  \
318    OK        40  53   1    NaN  119200  99968    57876     1      0   

        ...       rurlfarm  transprt  unemplyd  union   urban  whlretl  \
318     ...           1750     32948     27258   12.9  924952   117114   

     per_unemployed  per_age65  per_black  per_blucllr  
318        0.026003   0.113711   0.095365     0.055211  

[1 rows x 36 columns]

After doing some research on Oklahoma's first congressional district, they had republican congressman Steve Largent from 1994 to 2002. I will code that row as republican.

In [8]:
congress.repub.fillna(value = 1, inplace = True)

In [9]:
ind_vars = [ 'per_age65', 'per_black', 'per_blucllr', 'city','mdnincm', 'per_unemployed', 'union']
dep_var = 'repub'

x_const = sm.add_constant(congress[ind_vars])
y = congress[dep_var]

#Logit and probit models

logit_results = Logit(y, x_const).fit()

probit_results = Probit(y, x_const).fit()

Optimization terminated successfully.
         Current function value: 0.542084
         Iterations 7
Optimization terminated successfully.
         Current function value: 0.541259
         Iterations 7


Lets take a look at the summaries.

In [10]:
logit_results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  repub   No. Observations:                  435
Model:                          Logit   Df Residuals:                      427
Method:                           MLE   Df Model:                            7
Date:                Sat, 19 Nov 2016   Pseudo R-squ.:                  0.2166
Time:                        16:12:14   Log-Likelihood:                -235.81
converged:                       True   LL-Null:                       -301.01
                                        LLR p-value:                 5.158e-25
==================================================================================
                     coef    std err          z      P>|z|      [95.0% Conf. Int.]
----------------------------------------------------------------------------------
const              9.5500      1.618      5.902      0.000         6.378    12.722
per_age65        -11.9881      3.726     -3.217      0.001       -19.291    -4.685
per_black         -5.0358      1.273     -3.956      0.000        -7.531    -2.541
per_blucllr       -7.1161      6.269     -1.135      0.256       -19.404     5.171
city              -0.6513      0.259     -2.519      0.012        -1.158    -0.145
mdnincm        -5.843e-05   1.92e-05     -3.037      0.002     -9.61e-05 -2.07e-05
per_unemployed  -144.8827     23.332     -6.210      0.000      -190.613   -99.152
union             -0.0280      0.016     -1.705      0.088        -0.060     0.004
==================================================================================
"""

In [11]:
logit_marg = logit_results.get_margeff(dummy = True)
logit_marg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
        Logit Marginal Effects       
=====================================
Dep. Variable:                  repub
Method:                          dydx
At:                           overall
==================================================================================
                    dy/dx    std err          z      P>|z|      [95.0% Conf. Int.]
----------------------------------------------------------------------------------
per_age65         -2.2096      0.657     -3.363      0.001        -3.497    -0.922
per_black         -0.9282      0.220     -4.215      0.000        -1.360    -0.497
per_blucllr       -0.6093      0.092     -6.649      0.000        -0.789    -0.430
city              -0.1201      0.046     -2.586      0.010        -0.211    -0.029
mdnincm        -1.077e-05    3.4e-06     -3.163      0.002     -1.74e-05  -4.1e-06
per_unemployed   -26.7047      3.543     -7.538      0.000       -33.649   -19.761
union             -0.0052      0.003     -1.725      0.085        -0.011     0.001
==================================================================================
"""

Most notably, every 1 percent increase in unemployment decreases the probability of a republican congressman by 26.7%. Second to this, is the proportion of senior citizens: a 1 percent increase in the population of senior citizens correlates to a 2.2 % decrease in the probability of having a republican congressman.

Interestingly, the percentage of black population tracks at almost a -1:1 ratio with probability.

A 10,000 \$ increase in median income results in a (-1.0769172934225099e-05 * 10000) = -0.10769 % decrease in probability that that district elects a republican congressman

Now we look at probit summary and marginal effects

In [12]:
probit_results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                          Probit Regression Results                           
==============================================================================
Dep. Variable:                  repub   No. Observations:                  435
Model:                         Probit   Df Residuals:                      427
Method:                           MLE   Df Model:                            7
Date:                Sat, 19 Nov 2016   Pseudo R-squ.:                  0.2178
Time:                        16:12:14   Log-Likelihood:                -235.45
converged:                       True   LL-Null:                       -301.01
                                        LLR p-value:                 3.652e-25
==================================================================================
                     coef    std err          z      P>|z|      [95.0% Conf. Int.]
----------------------------------------------------------------------------------
const              5.7115      0.939      6.085      0.000         3.872     7.551
per_age65         -7.1058      2.184     -3.253      0.001       -11.387    -2.825
per_black         -3.0062      0.744     -4.038      0.000        -4.465    -1.547
per_blucllr       -4.2522      3.747     -1.135      0.256       -11.596     3.092
city              -0.3945      0.155     -2.541      0.011        -0.699    -0.090
mdnincm        -3.476e-05   1.15e-05     -3.024      0.002     -5.73e-05 -1.22e-05
per_unemployed   -86.8419     13.524     -6.421      0.000      -113.349   -60.335
union             -0.0171      0.010     -1.745      0.081        -0.036     0.002
==================================================================================
"""

In [13]:
probit_marg = probit_results.get_margeff()
probit_marg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
       Probit Marginal Effects       
=====================================
Dep. Variable:                  repub
Method:                          dydx
At:                           overall
==================================================================================
                    dy/dx    std err          z      P>|z|      [95.0% Conf. Int.]
----------------------------------------------------------------------------------
per_age65         -2.1788      0.646     -3.373      0.001        -3.445    -0.913
per_black         -0.9218      0.216     -4.273      0.000        -1.345    -0.499
per_blucllr       -1.3038      1.144     -1.140      0.254        -3.546     0.939
city              -0.1210      0.047     -2.599      0.009        -0.212    -0.030
mdnincm        -1.066e-05   3.42e-06     -3.121      0.002     -1.74e-05 -3.96e-06
per_unemployed   -26.6275      3.521     -7.563      0.000       -33.528   -19.727
union             -0.0052      0.003     -1.763      0.078        -0.011     0.001
==================================================================================
"""

In [14]:
np.abs(probit_marg.margeff) - np.abs(logit_marg.margeff)

array([ -3.08426013e-02,  -6.42455778e-03,   6.94486249e-01,
         9.23354580e-04,  -1.10433743e-07,  -7.71184393e-02,
         8.50164399e-05])

As you can see, the marginal effects of both probit and logit are almost identical, with small decimal differences between them.


Now, I show the OLS model:

In [15]:
ols_results = sm.OLS(y, x_const).fit()
ols_results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  repub   R-squared:                       0.224
Model:                            OLS   Adj. R-squared:                  0.211
Method:                 Least Squares   F-statistic:                     17.59
Date:                Sat, 19 Nov 2016   Prob (F-statistic):           1.63e-20
Time:                        16:12:14   Log-Likelihood:                -260.11
No. Observations:                 435   AIC:                             536.2
Df Residuals:                     427   BIC:                             568.8
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [95.0% Conf. Int.]
----------------------------------------------------------------------------------
const              2.0296      0.272      7.450      0.000         1.494     2.565
per_age65         -1.9612      0.689     -2.845      0.005        -3.316    -0.606
per_black         -0.4659      0.170     -2.741      0.006        -0.800    -0.132
per_blucllr       -1.6495      1.163     -1.418      0.157        -3.936     0.637
city              -0.1149      0.050     -2.307      0.022        -0.213    -0.017
mdnincm        -9.758e-06   3.58e-06     -2.728      0.007     -1.68e-05 -2.73e-06
per_unemployed   -22.5908      3.810     -5.929      0.000       -30.080   -15.101
union             -0.0027      0.003     -0.902      0.368        -0.009     0.003
==============================================================================
Omnibus:                        5.824   Durbin-Watson:                   1.820
Prob(Omnibus):                  0.054   Jarque-Bera (JB):               43.811
Skew:                          -0.284   Prob(JB):                     3.07e-10
Kurtosis:                       1.553   Cond. No.                     6.68e+06
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.68e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Just like the logit and probit models, the percentage of unemployed people has the biggest effect on the chances of a republican congressman. For every percentage point of unemployment, the chances of a republican go down by 22.6 %. The feature with the second biggest effect is the percentage of the population that is over age 65. For every percent increase in senior citizens, the chances of a republican decrease by 2 %.


### (b)

For the first 15 observations calculate $p(x_i,\beta)$ and whether the model predicts whether a district is republican or not 

In [16]:
from scipy.stats import norm

In [17]:
beta_matrix = np.matrix(probit_results.params.values )
beta_matrix = np.reshape(beta_matrix, (-1, 1))
x_matrix = np.matrix(x_const[:15])

In [18]:
xi_b = x_matrix * beta_matrix
p_xi_b = norm.cdf(xi_b)

In [19]:
by_hand = x_const.ix[:14, :]
by_hand['P(x_i, B)'] = pd.DataFrame(p_xi_b)

/Users/benjamindykstra/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app


In [20]:
by_hand

const  per_age65  per_black  per_blucllr  city  mdnincm  per_unemployed  \
0       1   0.041167   0.035870     0.050167     0    46581        0.041701   
1       1   0.125572   0.284881     0.075219     0    27360        0.034276   
2       1   0.130623   0.241355     0.086427     0    29492        0.024475   
3       1   0.129101   0.260216     0.114100     0    26800        0.029982   
4       1   0.146763   0.066002     0.128355     0    25401        0.028981   
5       1   0.112181   0.148305     0.084371     0    33189        0.026052   
6       1   0.121219   0.091056     0.052666     0    38768        0.018698   
7       1   0.139896   0.674458     0.083411     0    20773        0.041203   
8       1   0.151819   0.178819     0.095411     0    21889        0.032751   
9       1   0.123101   0.176122     0.073858     0    30011        0.025580   
10      1   0.156435   0.015736     0.100025     0    25935        0.021619   
11      1   0.163916   0.265669     0.089097     0    24217        0.029940   
12      1   0.091063   0.031659     0.051790     1    38232        0.027155   
13      1   0.094979   0.066459     0.069745     1    22650        0.045021   
14      1   0.194176   0.018221     0.050173     1    31663        0.025167   

        union  P(x_i, B)  
0   30.400000   0.253639  
1   18.200001   0.275506  
2   18.200001   0.590219  
3   18.200001   0.374448  
4   18.200001   0.584390  
5   18.200001   0.648832  
6   18.200001   0.857586  
7   18.200001   0.011405  
8   13.200000   0.443703  
9   13.200000   0.692401  
10  13.200000   0.869142  
11  13.200000   0.382629  
12  12.800000   0.672981  
13  12.800000   0.220520  
14  12.800000   0.565042

In [21]:
by_hand['predict_repub'] = by_hand['P(x_i, B)'].apply(lambda x: 1 if x >= 0.5 else 0)

/Users/benjamindykstra/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


In [22]:
by_hand['repub'] = y[:15]

/Users/benjamindykstra/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


Contrast predicted vote with the actual vote given by each of the first 15 individuals. The column 'predict_repub' are the predictions, and the 'repub' column are the actual results. The predictions are done at a 0.5 threshold: If the model predicts more than a 50% chance of republican, then the prediction is 1, else 0.

In [23]:
by_hand

const  per_age65  per_black  per_blucllr  city  mdnincm  per_unemployed  \
0       1   0.041167   0.035870     0.050167     0    46581        0.041701   
1       1   0.125572   0.284881     0.075219     0    27360        0.034276   
2       1   0.130623   0.241355     0.086427     0    29492        0.024475   
3       1   0.129101   0.260216     0.114100     0    26800        0.029982   
4       1   0.146763   0.066002     0.128355     0    25401        0.028981   
5       1   0.112181   0.148305     0.084371     0    33189        0.026052   
6       1   0.121219   0.091056     0.052666     0    38768        0.018698   
7       1   0.139896   0.674458     0.083411     0    20773        0.041203   
8       1   0.151819   0.178819     0.095411     0    21889        0.032751   
9       1   0.123101   0.176122     0.073858     0    30011        0.025580   
10      1   0.156435   0.015736     0.100025     0    25935        0.021619   
11      1   0.163916   0.265669     0.089097     0    24217        0.029940   
12      1   0.091063   0.031659     0.051790     1    38232        0.027155   
13      1   0.094979   0.066459     0.069745     1    22650        0.045021   
14      1   0.194176   0.018221     0.050173     1    31663        0.025167   

        union  P(x_i, B)  predict_repub  repub  
0   30.400000   0.253639              0    1.0  
1   18.200001   0.275506              0    1.0  
2   18.200001   0.590219              1    1.0  
3   18.200001   0.374448              0    1.0  
4   18.200001   0.584390              1    1.0  
5   18.200001   0.648832              1    0.0  
6   18.200001   0.857586              1    1.0  
7   18.200001   0.011405              0    0.0  
8   13.200000   0.443703              0    0.0  
9   13.200000   0.692401              1    0.0  
10  13.200000   0.869142              1    1.0  
11  13.200000   0.382629              0    1.0  
12  12.800000   0.672981              1    1.0  
13  12.800000   0.220520              0    0.0  
14  12.800000   0.565042              1    1.0

The confusion matrix has the the form

|      | Actual        |                | Total |
|-------------- | ------------- | -------------- | ----- |
|   Predicted   | 0             | 1              |       |
| 0             | true negative | false negative |row sum |
| 1             | false positive| true positive  |row sum |
| Total         | column sum    | column sum     |total sum |

In [24]:
from sklearn.metrics import confusion_matrix, classification_report

In [25]:
confusion_matrix(by_hand['repub'], by_hand['predict_repub'])

array([[3, 2],
       [4, 6]])

In [26]:
print classification_report(by_hand['repub'], by_hand['predict_repub'])

             precision    recall  f1-score   support

        0.0       0.43      0.60      0.50         5
        1.0       0.75      0.60      0.67        10

avg / total       0.64      0.60      0.61        15



So our model only classified democrats correctly in 3 / 7 times. It classified republicans correctly 6 / 8 times. What this tells me is that the model is pretty heavy handed in classifying republicans. I suspect that some factors have a greater influence on the probability than others, namely unemployment and the black population. 

### (c)

Now, I use the model to find where the Democrats should focus their resources. My criteria for this are districts where the probability of going republican is lower than 0.55 - 'swing districts' if you will. 

In [30]:
congress['probability'] = probit_results.predict()

congress[congress.probability.apply(lambda prob: True if (prob >= 0.5 and prob <= 0.55) else False)]

state  fipstate  sc  cd  repub   age65   black  blucllr  city  coast  \
19     CA         6  71   2    1.0   89325    8410    29098     0      0   
34     CA         6  71  17    0.0   59816   24991    30290     0      1   
42     CA         6  71  25    1.0   43152   25864    26610     1      0   
51     CA         6  71  34    0.0   50565   11212    55415     0      0   
66     CA         6  71  49    1.0   71244   30024    19415     1      0   
68     CA         6  71  51    1.0   65052    9333    24025     1      0   
103    FL        12  43  21    1.0   55885   23431    40625     1      0   
147    IN        18  22   2    1.0   75839   22634    56329     0      0   
164    KY        21  51   5    1.0   74683    5728    45791     0      0   
168    LA        22  45   3    1.0   89478  203766    73010     0      1   
185    MD        24  52   3    0.0   79329  103564    30591     1      1   
200    MI        26  23   8    0.0   52360   33396    40515     0      0   
203    MI        26  23  11    1.0   71694   23734    25715     0      0   
224    MO        29  34   8    1.0   92032   24940    54853     0      0   
226    MS        28  46   1    1.0   67906  117486    67369     0      0   
270    NY        36  13   3    1.0   78042   11417    22396     0      1   
303    OH        39  24   5    1.0   73877   12059    67381     0      1   
346    PA        42  14  18    0.0  104784   43784    27590     0      0   
349    PA        42  14  21    1.0   83625   21737    49211     0      0   
368    TX        48  49   1    0.0   90383  102299    44899     0      0   
381    TX        48  49  14    1.0   76103   60270    38531     1      1   
394    TX        48  49  27    0.0   57683   13771    28929     0      1   

        ...       transprt  unemplyd      union   urban  whlretl  \
19      ...           8040     18987  25.400000  185449    48240   
34      ...           7374     18759  25.400000  424528    54779   
42      ...           9213     14237  25.400000  506718    57774   
51      ...          13773     18726  25.400000  573456    58985   
66      ...           9186     14314  25.400000  573437    62206   
68      ...           8320     13278  25.400000  553964    61953   
103     ...          24786     18107   9.600000  557737    72140   
147     ...           8074     16693  25.100000  161951    55377   
164     ...          10253     25431  20.400000       0    41204   
168     ...          24195     33656  13.800000  350552    83487   
185     ...          12600     12348  18.600000  586569    61733   
200     ...           8017     16717  33.700001  291041    60577   
203     ...           7995     12518  33.700001  540673    69532   
224     ...          11581     18166  26.600000       0    46882   
226     ...          11347     14750   9.300000   29324    41371   
270     ...          16944     11630  35.799999  578040    65314   
303     ...           9944     16374  27.400000   31507    49877   
346     ...          12211     14576  27.000000  554352    62534   
349     ...           9783     16587  27.000000  224308    54220   
368     ...           8758     16462  12.500000   74553    48569   
381     ...           7750     15887  12.500000   55833    50022   
394     ...           8046     23087  12.500000  454818    50889   

     per_unemployed  per_age65  per_black  per_blucllr  probability  
19         0.033047   0.155472   0.014638     0.050646     0.504886  
34         0.032848   0.104742   0.043761     0.053040     0.509042  
42         0.024838   0.075284   0.045123     0.046424     0.530764  
51         0.032655   0.088176   0.019552     0.096633     0.504020  
66         0.024962   0.124240   0.052358     0.033857     0.549205  
68         0.023179   0.113559   0.016292     0.041939     0.533052  
103        0.032196   0.099368   0.041662     0.072235     0.507156  
147        0.030114   0.136814   0.040832     0.101618     0.517377  
164        0.040700   0.119524   0.009167     0.073285     0

### (d) 
For both probit and logit, discuss the predictive accuracy of the model by comparing the predicted outcomes with actual chosen outcomes.

In [43]:
# countR2 = # correctly predicted / total #

preds_probit = [1 if x >= 0.5 else 0 for x in probit_results.predict()]
preds_logit = [1 if x >= 0.5 else 0 for x in logit_results.predict()]

In [44]:
confusion_matrix(congress['repub'], preds_probit)

array([[139,  68],
       [ 49, 179]])

In [45]:
confusion_matrix(congress['repub'], preds_logit)

array([[139,  68],
       [ 50, 178]])

In [52]:
countR2_probit =  (confusion_matrix(congress['repub'], preds_probit)[0,0] + 
confusion_matrix(congress['repub'], preds_probit)[1,1]) / np.sum(confusion_matrix(congress['repub'], preds_probit))

print "the count R^2 for the probit model: ", countR2_probit

the count R^2 for the probit model:  0.731034482759


In [53]:
countR2_logit =  (confusion_matrix(congress['repub'], preds_logit)[0,0] + 
confusion_matrix(congress['repub'], preds_logit)[1,1]) / np.sum(confusion_matrix(congress['repub'], preds_logit))

print "the count R^2 for the logit model: ", countR2_logit

the count R^2 for the logit model:  0.728735632184


So far, probit and logit are neck and neck. Their classification accuracy is almost exactly the same for both models, with probit doing *slightly* better. Lets take a look at McFadden's R2 aka pseudo R2:

In [55]:
probit_results.prsquared, logit_results.prsquared

(0.21781295278040236, 0.21662053801812553)

Again, probit does slightly better than the logit. 

Lets dive in to what each measure of fit is capturing.

**Count R2** captures how well the model is classifying overall . It is not like other classic types of R2. It does not measure any kind of information gain from a null model to a fully specified model.

**McFadden's R2** = (1 - (log likelihood full model / log likelihood only intercept)). The log likelihood of the intercept model is treated as a total sum of squares, and the log likelihood of the full model is treated as the sum of squared errors (like in approach 1). The ratio of the likelihoods suggests the level of improvement over the intercept model offered by the full model. 


### (e) 
Which model do I recommend?
